### ISSUE TITLE
proxyForInterface should allow upcalls

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2901]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#2901|
|Type|enhancement|
|Created|2007-11-14 22:19:42Z|
|Branch|https://github.com/twisted/twisted/tree/proxyForInterface-upcall-2901-2|

Currently:

```python

class IFoo(Interface):
    def foo():
        pass

class TrivialDecorator(proxyForInterface(IFoo)):
    pass

class LessTrivialDecorator(TrivialDecorator):
    def foo(self):
        return TrivialDecorator.foo(self)

LessTrivialDecorator(None).foo()
```

will fail with

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 3, in foo
TypeError: 'property' object is not callable
```

This is a problem because it means that subclasses of subclasses of `proxyForInterface` classes will need to do different things to access parent class behaviour, depending on which methods their parent overrode.

<details><summary>Searchable metadata</summary>

```
trac-id__2901 2901
type__enhancement enhancement
reporter__jml jml
priority__highest highest
milestone__ 
branch__branches_proxyForInterface_upcall_2901_2 branches/proxyForInterface-upcall-2901-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1195078782000000 1195078782000000
changetime__1196281303000000 1196281303000000
version__None None
owner__ 
cc__radix cc__glyph cc__jml cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

This should be done before release so we aren't lugging around another deprecated API.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

I gave a try at this, and it doesn't seem very easy. I'd like to propose a change in the workflow here: first create a branch with a new TODO test, with a reference to this ticket. And then try to find a solution.

Of course, it's only valuable if there's not an obvious solution :).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

Personally I don't think it's worth making such a branch.

radix, glyph: Got suggestions on how to solve this?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

In my experience, once one has a failing TODO test, the rest is easy, and there's no point in creating a second branch :).

I'll have a look at how to solve it, though...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

(In [21968]) Branching to 'proxyForInterface-upcall-2901'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner|
|-|-|

Ready for review.

I did it really fast, so maybe I made some mistakes.  On the other hand, it was really easy, so maybe I didn't.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

(In [21970]) re [#2901](https://github.com/twisted/twisted/issues/2901) jml noted that my usage of new-style classes was inconsistent.  not a review comment technically, but nevertheless...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set owner to @glyph|
|-|-|

Damn I wasn't very far from the solution :). Good job.

Things I've spotted:
 * _proxiedClassMethod and _proxyDescriptor should probably be named _ProxiedClassMethod and _ProxyDescriptor
 * attributeName of _proxyDescriptor is undocumented
 * there is probably a missing test for checking that arguments are correctly pass to the ```_proxiedClassMethod.__call__ method``` (if I change the signature of ```__call__``` nothing breaks).
 * not new in this branch, but instead of doing ```for name, description in iface.namesAndDescriptions()```, you can probably do ```for name in iface.names()```. Or use the description value :).

This is going to conflict pretty badly with [#2895](https://github.com/twisted/twisted/issues/2895), so this ticket should probably go in first.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @glyph|
|-|-|

Perfect. Please limit the docstring of _ProxiedClassMethod to 80 columns, then merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set status to `closed`|
|-|-|

(In [21987]) Correct the behavior of `proxyForInterface` with respect to inheritance.

Author: glyph

Reviewer: therve

Fixes [#2901](https://github.com/twisted/twisted/issues/2901)

This corrects two bugs in `proxyForInterface`.  The first is that subclasses of types generated for `proxyForInterface` were not providing convincing facimilies of the methods that they allegedly implemented; you could not define a proxy class `Foo` and a subclass `Bar`, and have `Bar` call `Foo.method(self)`.  Now you can.

The other is that proxyies for sub-interfaces were incorrectly restricting access to the methods of their superclasses.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [21988]) Revert r21987: missing tests for __get__ code path.

Refs [#2901](https://github.com/twisted/twisted/issues/2901)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set status to `reopened`|
|-|-|

There is a missing test for type=None in get property.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

(In [21990]) Branching to 'proxyForInterface-upcall-2901-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

(In [21992]) re [#2901](https://github.com/twisted/twisted/issues/2901) do it right

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

(In [21993]) re [#2901](https://github.com/twisted/twisted/issues/2901) test coverage for that case.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

Test added, changeset in previous comment.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_20"></a>|@therve set owner to @glyph|
|-|-|

I tried hard to find something to say, but didn't manage to find anything. Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set status to `closed`|
|-|-|

(In [21994]) Correct the behavior of `proxyForInterface` with respect to inheritance.

(This is a repeat of r21987 with an additional test.)

Author: glyph

Reviewer: therve

Fixes [#2901](https://github.com/twisted/twisted/issues/2901)

This corrects two bugs in `proxyForInterface`.  The first is that subclasses of types generated for `proxyForInterface` were not providing c
onvincing facimilies of the methods that they allegedly implemented; you could not define a proxy class `Foo` and a subclass `Bar`, and have
 `Bar` call `Foo.method(self)`.  Now you can.

The other is that proxyies for sub-interfaces were incorrectly restricting access to the methods of their superclasses.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
